### PR TITLE
[Feature branch] Lower bounds for min-max normalization in hybrid query

### PR DESCRIPTION
**Pull request by martin-gaievski**
### Description

I'm adding `lower_bounds` parameter for min-max normalization technique of hybrid query.
User will be able to create search pipeline with `lower_bounds` using following request:

```
{
  "description": "Normalization processor for hybrid search",
  "phase_results_processors": [
    {
      "normalization-processor": {
        "normalization": {
          "technique": "min_max",
          "parameters": {
            "lower_bounds": [
              { 
                "mode": "apply",
                "min_score": 0.1
              }, 
                "mode": "clip",
                "min_score": 0.1
              }, 
                "mode": "ignore"
              }
            ]
          }
        },
        "combination": {
          "technique": "arithmetic_mean"
        }
      }
    }
  ]
}
```

I'm planning to merge this into feature branch in main repo until app sec team give a signoff. 

### Related Issues
https://github.com/opensearch-project/neural-search/issues/150
Implemented design described in RFC https://github.com/opensearch-project/neural-search/issues/1189

### Check List
- [X] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [X] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/neural-search/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java [hunk 1]
```diff
@@ -57,7 +57,7 @@ public final class HybridQueryBuilder extends AbstractQueryBuilder<HybridQueryBu
 
     private Integer paginationDepth;
 
-    static final int MAX_NUMBER_OF_SUB_QUERIES = 5;
+    public static final int MAX_NUMBER_OF_SUB_QUERIES = 5;
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:60]: making this public to reaad and make validation in other places down the stream

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 2]
```diff
@@ -167,22 +225,182 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
+                    Locale.ROOT,
+                    "lower_bounds size %d should be less than or equal to %d",
+                    lowerBoundsParams.size(),
+                    MAX_NUMBER_OF_SUB_QUERIES
+                )
+            );
+        }
+
+        for (Object boundObj : lowerBoundsParams) {
+            if (!(boundObj instanceof Map)) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:263]:
```suggestion
            if ((boundObj instanceof Map)==false) {
```
improves readability

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 3]
```diff
@@ -167,22 +225,182 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
+
+    private Optional<List<Pair<LowerBound.Mode, Float>>> getLowerBounds(final Map<String, Object> params) {
+        if (Objects.isNull(params) || !params.containsKey(PARAM_NAME_LOWER_BOUNDS)) {
+            return Optional.empty();
+        }
+
+        List<Pair<LowerBound.Mode, Float>> lowerBounds = new ArrayList<>();
+
+        Object lowerBoundsObj = params.get(PARAM_NAME_LOWER_BOUNDS);
+        if (!(lowerBoundsObj instanceof List<?> lowerBoundsParams)) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:247]:
```suggestion
        if ((lowerBoundsObj instanceof List<?> lowerBoundsParams)==false) {
```

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 4]
```diff
@@ -167,22 +225,182 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
 
-    private float normalizeSingleScore(final float score, final float minScore, final float maxScore) {
+    private float normalizeSingleScore(final float score, final float minScore, final float maxScore, final LowerBound lowerBound) {
         // edge case when there is only one score and min and max scores are same
         if (Floats.compare(maxScore, minScore) == 0 && Floats.compare(maxScore, score) == 0) {
             return SINGLE_RESULT_SCORE;
         }
-        float normalizedScore = (score - minScore) / (maxScore - minScore);
-        return normalizedScore == 0.0f ? MIN_SCORE : normalizedScore;
+        if (!lowerBound.isEnabled()) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:233]:
```suggestion
        if (lowerBound.isEnabled()==false) {
```

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 5]
```diff
@@ -167,22 +225,182 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
+            }
+
+            @SuppressWarnings("unchecked")
+            Map<String, Object> lowerBound = (Map<String, Object>) boundObj;
+
+            try {
+                LowerBound.Mode mode = LowerBound.Mode.fromString(
+                    Objects.isNull(lowerBound.get(PARAM_NAME_LOWER_BOUND_MODE))
+                        ? ""
+                        : lowerBound.get(PARAM_NAME_LOWER_BOUND_MODE).toString()
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:274]:
```suggestion
                LowerBound.Mode mode = LowerBound.Mode.fromString(
                    Objects.nonNull(lowerBound.get(PARAM_NAME_LOWER_BOUND_MODE))
                        ? lowerBound.get(PARAM_NAME_LOWER_BOUND_MODE).toString():""
```

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 6]
```diff
@@ -167,22 +225,182 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
         }
-        float normalizedScore = (score - minScore) / (maxScore - minScore);
-        return normalizedScore == 0.0f ? MIN_SCORE : normalizedScore;
+        if (!lowerBound.isEnabled()) {
+            return LowerBound.Mode.IGNORE.normalize(score, minScore, maxScore, lowerBound.getMinScore());
+        }
+        return lowerBound.getMode().normalize(score, minScore, maxScore, lowerBound.getMinScore());
+    }
+
+    private Optional<List<Pair<LowerBound.Mode, Float>>> getLowerBounds(final Map<String, Object> params) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:239]: Please add comments on what is happening in this method to better understand.

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 7]
```diff
@@ -34,8 +42,28 @@
 public class MinMaxScoreNormalizationTechnique implements ScoreNormalizationTechnique, ExplainableTechnique {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:42]: We have added static class LowerBound in the same class. Can we extract it to independent file? If in future we add similar functionality then we have to rework to extract Lower bound and put it in another file.

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationUtil.java [hunk 8]
```diff
@@ -82,4 +82,78 @@ public static void setNormalizedScore(
...
+    ) {
+        if (Objects.isNull(actualParameters) || actualParameters.isEmpty()) {
+            return;
+        }
+        boolean hasUnknownParameters = false;
+        for (Map.Entry<String, Object> entry : actualParameters.entrySet()) {
+            String paramName = entry.getKey();
+            Object paramValue = entry.getValue();
+
+            if (!supportedParametersTopLevel.contains(paramName)) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationUtil.java:115]:
```suggestion
            if (supportedParametersTopLevel.contains(paramName)==false) {
```

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 4 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:233]: It is a nit. Just for better readablity

**Review by vibrantvarun — COMMENTED** [review]
Completed 1st round of review

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 7 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:42]: not sure I got your reasoning. I don't think it needs to be in a separate file, it's essentially a record class, and I don't see a need for it in other techniques in near future. We can do this later if needed, currently it's over engineering

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 4 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:233]: I'm not in favor of this style, will keep it unchanged

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:239]: yes, I think I can split into two mwethods

Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:239]: ack

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 3 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:247]: same as above, I do like the minimalistic syntax

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 2 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:263]: same as above, I do like the minimalistic syntax

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 5 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:274]: not sure what's the benefit of this change, and format is enforced by spotless, will keep my version

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationUtil.java (hunk 8 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationUtil.java:115]: same as above, I do like the minimalistic syntax

File: src/main/java/org/opensearch/neuralsearch/processor/CompoundTopDocs.java [hunk 9]
```diff
@@ -150,4 +151,78 @@ private ScoreDoc deepCopyScoreDoc(final ScoreDoc scoreDoc, final boolean isSortE
...
+            if ((firstDoc == null) != (secondDoc == null)) {
+                return false;
+            }
+            if (firstDoc == null) {
+                continue;
+            }
+            if (firstDoc.doc != secondDoc.doc || Float.compare(firstDoc.score, secondDoc.score) != 0) {
+                return false;
+            }
+            if (firstDoc instanceof FieldDoc != secondDoc instanceof FieldDoc) {
```
[the remark this answers is not in the document]
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/CompoundTopDocs.java:200]: I not sure why that's needed

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 2 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java]: @martin-gaievski checking boolean expression with `==` flag is [OpenSearch dev guide lines](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#java-language-formatting-guidelines). We should follow the same

Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java]: ok, let's follow the guideline in core then, although I don't see benefits in that syntax. I want to tweak plugin rules so this is applied automatically, will work on this in a separate issue: https://github.com/opensearch-project/neural-search/issues/1203

Comment (owaiskazi19) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java]: The above change won't be handled by spotless though. Spotless is used for styling and formatting and the above one is a code change

Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java]: to me this is styling, and this can be done via spotless custom rules

Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java]: I've changed all negations to explicit comparisons with false, should be good for changes in this PR

**Review by vibrantvarun — APPROVED** [review]
Looks good to me.

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 10]
```diff
@@ -167,22 +224,197 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
+        // validate that the input parameters are in correct format
+        if (Objects.isNull(params) || params.containsKey(PARAM_NAME_LOWER_BOUNDS) == false) {
+            return Optional.empty();
+        }
+
+        Object lowerBoundsObj = params.get(PARAM_NAME_LOWER_BOUNDS);
+        if (lowerBoundsObj instanceof List<?> == false) {
+            throw new IllegalArgumentException("lower_bounds must be a List");
+        }
+        List<?> lowerBoundsParams = (List<?>) lowerBoundsObj;
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:253]: How about?
```suggestion
      List<?> lowerBoundsParams = Optional.ofNullable(params.get(PARAM_NAME_LOWER_BOUNDS))
        .filter(List.class::isInstance)
        .map(List.class::cast)
        .orElseThrow(() -> new IllegalArgumentException("lower_bounds must be a non-null List"));
```

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 11]
```diff
@@ -167,22 +224,197 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
+        // number of lower bounds must match the number of sub-queries in a hybrid query
+        if (lowerBoundsParams.size() > MAX_NUMBER_OF_SUB_QUERIES) {
+            throw new IllegalArgumentException(
+                String.format(
+                    Locale.ROOT,
+                    "lower_bounds size %d should be less than or equal to %d",
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:259]: For my understanding
Don't we have to support lower bound for every sub query? For the ones we don't we can have mode as `ignore`. That should mean the lower bound size should be equal to sub query size.

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 12]
```diff
@@ -167,22 +224,197 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
+    private static class MinMaxScores {
         float[] minScoresPerSubquery;
         float[] maxScoresPerSubquery;
     }
+
+    /**
+     * Result class to hold lower bound for each sub query
+     */
+    @Getter
+    public static class LowerBound {
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:327]: Since this is a public class anyway. We could move it to a separate class file

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 13]
```diff
@@ -167,22 +224,197 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
+        static final float MIN_LOWER_BOUND_SCORE = -10_000f;
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:328]: How are deciding this bound scores for min and max?

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 14]
```diff
@@ -167,22 +224,197 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
+        /**
+         * Enum for normalization mode
+         */
+        protected enum Mode {
+            APPLY {
+                @Override
+                public float normalize(float score, float minScore, float maxScore, float lowerBoundScore) {
+                    // if we apply the lower bound this mean we use actual score in case it's less then the lower bound min score
+                    // same applied to case when actual max_score is less than lower bound min score
+                    if (maxScore < lowerBoundScore || score < lowerBoundScore) {
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:355]: Why are we checking `maxScore` here? Shouldn't it be `minScore` that needs to be compared with `lowerBoundScore`

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 15]
```diff
@@ -167,22 +224,197 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
+                    if (score < minScore) {
+                        return 0.0f;
+                    }
+                    if (maxScore < lowerBoundScore) {
+                        return (score - minScore) / (maxScore - minScore);
+                    }
+                    return (score - lowerBoundScore) / (maxScore - lowerBoundScore);
+                }
+            },
+            IGNORE {
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:375]: Do we need a separate enum for the original logic for min max? When the mode is IGNORE we could rely on the original normalize method of MinMax class

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 16]
```diff
@@ -167,22 +224,197 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
+        String lowerBoundModeValue = Objects.toString(lowerBound.get(PARAM_NAME_LOWER_BOUND_MODE), "");
+        LowerBound.Mode mode = LowerBound.Mode.fromString(lowerBoundModeValue);
+        float minScore = extractAndValidateMinScore(lowerBound);
+
+        return ImmutablePair.of(mode, minScore);
+    }
+
+    private float extractAndValidateMinScore(Map<String, Object> lowerBound) {
+        Object minScoreObj = lowerBound.get(PARAM_NAME_LOWER_BOUND_MIN_SCORE);
+        if (minScoreObj == null) {
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:293]: A question here. If a user is defining lower bounds then shouldn't we mandate them to provide a min score otherwise is there any point to define lower bound?

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 17]
```diff
@@ -167,22 +224,197 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
+            return LowerBound.DEFAULT_LOWER_BOUND_SCORE;
+        }
+        try {
+            float minScore = LowerBound.DEFAULT_LOWER_BOUND_SCORE;
+            if (Objects.nonNull(lowerBound.get(PARAM_NAME_LOWER_BOUND_MIN_SCORE))) {
+                minScore = Float.parseFloat(String.valueOf(lowerBound.get(PARAM_NAME_LOWER_BOUND_MIN_SCORE)));
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:299]: Should we also check if the provided min score is float if not throw an error?

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 18]
```diff
@@ -167,22 +224,197 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
+            },
+            CLIP {
+                @Override
+                public float normalize(float score, float minScore, float maxScore, float lowerBoundScore) {
+                    // apply clipping, return lower bound min score if score is less than min score. This effectively means 0 after
+                    // normalization
+                    if (score < minScore) {
+                        return 0.0f;
+                    }
+                    if (maxScore < lowerBoundScore) {
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:369]: Same here

**Review by owaiskazi19 — COMMENTED** [review]
Took a look except the test file and rest LGTM

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/L2ScoreNormalizationTechnique.java [hunk 19]
```diff
@@ -32,6 +33,14 @@ public class L2ScoreNormalizationTechnique implements ScoreNormalizationTechniqu
     public static final String TECHNIQUE_NAME = "l2";
     private static final float MIN_SCORE = 0.0f;
 
+    public L2ScoreNormalizationTechnique() {
+        this(Map.of(), new ScoreNormalizationUtil());
+    }
+
+    public L2ScoreNormalizationTechnique(final Map<String, Object> params, final ScoreNormalizationUtil scoreNormalizationUtil) {
+        scoreNormalizationUtil.validateParameters(params, Set.of(), Map.of());
+    }
```
[the remark this answers is not in the document]
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/L2ScoreNormalizationTechnique.java:42]: we actually validate, `params` has the actual parameters provided by user, and other two parameters mean we don't support any of those in L2 method. So if user actually sets some parameters this validation will fail with "your parameters are not supported".

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 20]
```diff
@@ -34,8 +41,28 @@
 public class MinMaxScoreNormalizationTechnique implements ScoreNormalizationTechnique, ExplainableTechnique {
     @ToString.Include
     public static final String TECHNIQUE_NAME = "min_max";
-    private static final float MIN_SCORE = 0.001f;
+    protected static final float MIN_SCORE = 0.001f;
     private static final float SINGLE_RESULT_SCORE = 1.0f;
+    private static final String PARAM_NAME_LOWER_BOUNDS = "lower_bounds";
+    private static final String PARAM_NAME_LOWER_BOUND_MODE = "mode";
+    private static final String PARAM_NAME_LOWER_BOUND_MIN_SCORE = "min_score";
```
[the remark this answers is not in the document]
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:48]: nope, each class has it's own set of constants, they are not common between techniques/classes

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 10 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:253]: ack will make this change

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 11 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:259]: yes, lower bound is for every sub-query. Here we're cutting extremely large collection with lower bounds even before we start parsing them. For hybrid query we can have up to 5 sub-queries, and here we're checking if lower bounds list does not have more then 5 elements.

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 12 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:327]: Public scope is too broad for this class, will change it to package private and keep it under MinMaxTechnique main class

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 13 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:328]: those are random numbers, just to have some limits to avoid extreme edge cases. In theory score are unbounded for base scoring (TF-IDF, BM25), but typically range between 0 and around 20-30 for most practical cases. For semantic queries it depends, for non-normalized vectors and spaces like inner product theory score are unbounded, but if vectors are normalized they are within [-1...1]

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 15 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:375]: we could for sure, but as you mentioned we're adding "ignore" as one of the new modes. What I can do is to refactor this logic into a common method that is used from both places: when not lower bound provided, and for "ignore" mode

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 16 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:293]: I think that was one of the asks when we reviewed the design/RFC, we can use default of 0.0 that is giving reasonable results and simplify the interface for end user

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 17 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:299]: this is checked in the wrapping try/catch, we do following exception in such case:
```
throw new IllegalArgumentException("invalid format for min_score: must be a valid float value", e);
```

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 14 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:355]: this is an edge case I bumped into while working on implementation. Imagine most of your docs are not very relevant to the query and the lower bound min_score is relatively high - this is the scenario we're catching here. We effectively falling back to traditional min-max for the lack of better option.

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 18 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:369]: please check my previous response, it's the case when user has the lower bound that is too high

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 15 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:375]: we already going just that, it no lower bounds we do call Mode.IGNORE.normalize() https://github.com/martin-gaievski/neural-search/blob/feature/lower_bounds_for_minmax_normalization/src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java#L232-L233

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/L2ScoreNormalizationTechnique.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/L2ScoreNormalizationTechnique.java:42]: we have a test for exactly this scenario https://github.com/martin-gaievski/neural-search/blob/feature/lower_bounds_for_minmax_normalization/src/test/java/org/opensearch/neuralsearch/processor/normalization/L2ScoreNormalizationTechniqueTests.java#L324

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java [hunk 21]
```diff
@@ -167,22 +224,196 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
...
+                        return (score - minScore) / (maxScore - minScore);
+                    }
+                    return (score - lowerBoundScore) / (maxScore - lowerBoundScore);
+                }
+            },
+            IGNORE {
+                @Override
+                public float normalize(float score, float minScore, float maxScore, float lowerBoundScore) {
+                    // ignore lower bound logic and do raw min-max normalization using actual scores
+                    float normalizedScore = (score - minScore) / (maxScore - minScore);
```
Comment (junqiu-lei) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:378]: How does the normalization logic handle negative scores? Are there edge cases when scores are negative, and have we added tests to cover these scenarios?

File: src/test/java/org/opensearch/neuralsearch/query/HybridQueryExplainIT.java [hunk 22]
```diff
@@ -703,6 +668,154 @@ public void testExplain_whenRRFProcessor_thenSuccessful() {
         assertTrue((double) explanationsHit4.get("value") > 0.0f);
     }
 
+    @SneakyThrows
+    public void testExplain_whenMinMaxNormalizationWithLowerBounds_thenSuccessful() {
```
Comment (junqiu-lei) [inline src/test/java/org/opensearch/neuralsearch/query/HybridQueryExplainIT.java:672]: Shall we also add some edge cases in IT, like case: lower bound greater than max score?

**Review by junqiu-lei — APPROVED** [review]
Overall LGTM.

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 21 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:378]: this is "ignore" mode that is the same logic we do have today for min-max. in this case there is no lower bound score involved, so none of these scores should be negative because:
- all scores are positive
- min and max calculated based on those positive scores

File: src/test/java/org/opensearch/neuralsearch/query/HybridQueryExplainIT.java (hunk 22 above)
Comment (martin-gaievski) [reply src/test/java/org/opensearch/neuralsearch/query/HybridQueryExplainIT.java:672]: added test case to org.opensearch.neuralsearch.processor.NormalizationProcessorIT

**Review by owaiskazi19 — COMMENTED** [review]
Looks good overall with few questions

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 13 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:327]: [-5. 5] can be too tight, even for typical TF-IDF, BM25 cases it can go as high as 30

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 16 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:292]: not sure what do you mean. I was referring to scenario when user sets just lower bound, in this case we use 0.0 score as min_score for that lower bound. Reason for it is a simpler request syntax and relatively good defaults for entry level users. User can be not aware of optimal min_score for lower bound, or can be fine with our defaults.

Comment (owaiskazi19) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:292]: So in the scenario where user just define a lower bound but doesn't pass any min score then it will be directed to default value, right?
Wouldn't in such scenario rather than using the default min value we should have used the original min score we get from the sub query itself?

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java (hunk 14 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java:354]: This should be common case, minScore should be lower then the lowerBound score in general. In this case we mainly case how actual score compares with lowerBound score, not minScore with lowerBound score. In case doc score is >= then lowerBound score we use lowerBound as min score:

```(score - lowerBoundScore) / (maxScore - lowerBoundScore);```

in case doc score < lowerBound score then we keep raw min-max formula:

```(score - minScore) / (maxScore - minScore)```

**Review by owaiskazi19 — APPROVED** [review]
